### ISSUE TITLE
Add breaks.txt

### ISSUE THREAD

**Issue by zaoqi** [minetest/minetest#5187]

**Comment by zaoqi** [comment]
mods
|-- modname
|   |-- breaks.txt
|   |-- depends.txt
|   |-- init.lua
|   |-- textures
|   |   |-- modname_stuff.png
|   |   \-- modname_something_else.png
|   |-- sounds
|   |-- media
|   \-- custom_data
|
\-- another_mod

**Comment by numberZero** [comment]
@zaoqi nice description) Do you mean a list of incompatible mods?

**Comment by sofar** [comment]
We've had several discussions about this topic already, and the consensus is to make `conflicts.txt` files instead. However, no merged patch has come out of this yet.

**Comment by DS-Minetest** [comment]
The conflict mods could be added to `depends.txt` with a suffix, like `blamod~`? I don't see a good reason to add an extra file for conflicting mods. Also if someone looks into depends.txt, he would see instantly, what mods are needed and which are disallowed.

**Comment by rubenwardy** [comment]
should be added to mod.conf instead
depends.txt should be deprecated

**Comment by sofar** [comment]
I agree with rubenwardy that instead of adding more files, we should try to unify them to `mod.conf` instead.

**Comment by rubenwardy** [comment]
There was a PR for this, iirc

**Comment by dsohler** [comment]
@sofar is right. Better reduce the amount of files needed for a mod. So Add it to `mod.conf` instead (and heavily extend the file’s capabilities)

~~~~
[meta]
name = my mod
author = me

[dependencies]
depends = modname1,modname2,modnameN
conflicts = modname1,modname2,modnameN
suggests = modname1,modname2,modnameN

[_costom]
_custom_attribute = custom value
~~~~

**Comment by numberZero** [comment]
perhaps add some version support? like `depends=libdoors:2` (>=2.0.0, <3.0.0)
also note that that’s not `conf`, in the sense it should not look like something user-configurable. there should be better places for that.

**Comment by rubenwardy** [comment]
> also note that that’s not conf, in the sense it should not look like something user-configurable. there should be better places for that.

The `.conf` is more the format (of `key = value`) than the content (imo)

**Comment by numberZero** [comment]
@rubenwardy 1. Are you still on Win32? 2. Are you aware of that `conf` is short for `configuration`?
Seriously, `world.mt` seems to use the same format. And some files have no extension at all. Also `.conf` is used by some programs (e.g. Apache HTTPD) for a different format...
The problem is that a mod might have real configuration file, and that would be confusing. The `mod.conf` file is not a mod configuration, it’s rather mod description/specification/metainformation.,. oh, what’s about `mod.meta`?

**Comment by rubenwardy** [comment]
1. No, not for years. 2. Yes, so?

**Comment by red-001** [comment]
Other programs also using the extension seems hardly relevant.

**Comment by dsohler** [comment]
The file suffix is one of the least relevant things here.

**Comment by numberZero** [comment]
Sorry for that, I was probably too sleepy.
But anyway, what’s about versioning?

**Comment by dsohler** [comment]
`modname:version` as you suggested seems to be a good idea.

**Comment by DS-Minetest** [comment]
But then the version should be only suggested, not forced. Most mods depend not on only one version of another mod.

**Comment by DS-Minetest** [comment]
Also sometimes there are some mods with the same name eg. `nether`. This can be confusing.

**Comment by dsohler** [comment]
Yeah, there are always made-up edge cases that prevent developers to implement something because it could cause "too much complexity". But those edge cases are what they are: edge cases.

Mods should be able to "force" a specific version if they depend on that version (or force a minimum version using `modname:>=version`) but users should be able to configure that the mod should be loaded anyways (or get a prompt telling the user that running mod X requires mod Y in version >=N).

**Comment by DS-Minetest** [comment]
Better than a minimum would be some sort of interval (`]version1;version2]` or `[version2;inf[`) because often newer versions of mods don't work anymore.

**Comment by dsohler** [comment]
As I said: It’s easy to make up edge cases (here: a mod that needs another mod above version X but below version Y). But there should at least all comparison operators be available and usable (`<`, `>`, `=`, `^`, `$`, `~` and `!` in all common combinations).

**Comment by red-001** [comment]
seems complicated, whats the use case? is there a lot of mods that you need a specific version of? Seems kinda like an edge case. I could be wrong of course.

**Comment by dsohler** [comment]
I guess when implementing a simple comparison of numbers using `<` and `>` adding `^`, `$`, `~` and `!` wouldn't be that much of an issue since all modern relevant languages already have them (at least `!`).

But yes, being able to use `^`, `$` and `~` would be fun but actually isn't really needed.

**Comment by ShadowNinja** [comment]
As of now, Minetest doesn't have a concept of mod versions.  However, if it is added, we should use a standard version selection system like that offered by pip (Python) or Cargo (Rust)<sup>[1]</sup>.  We should also encourage semver.

[1]: pip uses a `mod>=1.2.3` style, cargo uses a `mod = ">=1.2.3"` style which could be implemented in Minetest as:
```
dependencies = {
        mod = "^1.2.3"
}
```

For now, #4539 is the PR to watch for this.  It uses a space-seperated system that would work best with the pip style.

**Comment by sfan5** [comment]
As touched upon here https://github.com/luanti-org/luanti/issues/15706#issuecomment-2613164945 I think it's reasonable to add `conflicts` so CDB can make the right suggestions, but please don't implement more complicated relationships.
